src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -246,17 +320,24 @@ static AppInstallationToken generateAppInstallationToken(
                 appInstallation = appInstallationOptional.get();
             }
 
-            GHAppInstallationToken appInstallationToken = appInstallation
-                    .createToken(appInstallation.getPermissions())
-                    .create();
+            GHAppCreateTokenBuilder builder = appInstallation.createToken();
+            if (!repositories.isEmpty()) {
+                builder.repositories(repositories);
+            }
+            if (!permissions.isEmpty()) {
+                builder.permissions(permissions);
+            } else {
+                builder.permissions(appInstallation.getPermissions());
+            }
+            GHAppInstallationToken appInstallationToken = builder.create();
```

Comment:
These new methods from https://github.com/hub4j/github-api are what actually change the accessible repositories and permissions for the generated tokens.